### PR TITLE
[WIP] Fix invoking a global function pointer defined using var

### LINKED ISSUE #18979: Can't invoke a global function pointer defined using var
**Description:**
$title. Works for local variables.

**Steps to reproduce:**
```ballerina
import ballerina/io;

var f1 = function() returns string {
    return "f1";
};

public function main() {
    var f2 = function() returns string {
        return "hi";
    };

    // string s1 = f1(); // <- FAILS undefined function 'f1'
    string s2 = f2();

    io:println(s2);
}
```

**Affected Versions:**
```cmd
$ ballerina -v
Ballerina 1.0.0-rc1-SNAPSHOT
Language specification 2019R3
```

### PR DESCRIPTION
**Pull request by irshadnilam**
## Purpose
$title

Fixes #18979

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 1]
```diff
@@ -304,6 +304,10 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
         pkgNode.typeDefinitions.forEach(typDef -> typDefs.add(typDef));
         defineTypeNodes(typDefs, pkgEnv);
 
+        pkgNode.globalVars.stream().filter(variable -> variable.expr != null &&
+                variable.expr.getKind() == NodeKind.LAMBDA)
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Shall we split the line before `&&`

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/lambda/function-pointers.bal [hunk 2]
```diff
@@ -159,3 +159,11 @@ function testSubTypingWithAny() returns any {
     any s = ff();
     return s;
 }
+
+var f1 = function() returns string {
+    return "f1";
+};
+
+public function testGlobalFunctionPointerVar() returns string{
```
Comment (rdhananjaya) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/lambda/function-pointers.bal]: Need a space before `{'

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 3]
```diff
@@ -304,6 +304,10 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
         pkgNode.typeDefinitions.forEach(typDef -> typDefs.add(typDef));
         defineTypeNodes(typDefs, pkgEnv);
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Shall we add a comment summarizing the reason for this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 1 above)
Comment (irshadnilam) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java (hunk 3 above)
Comment (irshadnilam) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Added.
`Adding type info for global functions that are declared using var`
Would this comment be okay?

Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: How about specifying why we do this. Something like `Types of module level function pointers declared with 'var' are resolved before referencing them in functions as TypeNode is null for them.`

Comment (irshadnilam) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Fixed. Thanks

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 4]
```diff
@@ -667,6 +673,18 @@ private void checkErrors(TypeDefinition unresolvedType, BLangType currentTypeNod
         }
     }
 
+    private void addFunctionType(BLangSimpleVariable varNode, SymbolEnv env) {
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Can't we re-use an existing method? This doesn't seem to handle rest parameters.

**Comment by irshadnilam** [comment]
This fix does not solve globally defined functions with rest parameters because of `BInvokableType` not having a field for restparams.  I'll create a seperate issue to track that. 

Closing this.

**Pull request closed without merging**